### ISSUE TITLE
[core] Scaling issues on Wayland

### ISSUE THREAD

**Issue by sagehane** [raysan5/raylib#2566]
**WARNING: Please, read this note carefully before submitting a new issue:**

It is important to realise that **this is NOT A SUPPORT FORUM**, this is for reproducible BUGS with raylib ONLY.

There are lots of generous and helpful people ready to help you out on [raylib Discord forum](https://discord.gg/raylib) or [raylib reddit](https://www.reddit.com/r/raylib/).

Remember that asking for support questions here actively takes developer time away from improving raylib.

---

Please, before submitting a new issue verify and check:

 - [x] I tested it on latest raylib version from master branch
 - [x] I checked there is no similar issue already reported
 - [x] My code has no errors or misuse of raylib

### Issue description

Note: Likely because of an upstream bug (https://github.com/glfw/glfw/issues/1506).
I wasn't sure so I decided to make an issue here anyways. Feel free to close if unactionable on Raylib's side.

Most things don't seem to scale properly when running Raylib with Wayland at a higher scaling.
Some of the stuff that seem to scale properly are screen dimensions and background color.
This issue doesn't occur on the X11 version.

Tested on versions 3.3.7 and the latest commit.

### Environment

Platform/OS: NixOS Unstable
```
$ nixos-version
22.11pre388782.cf3ab54b4af (Raccoon)
```

Window Manager: Master branch of [RiverWM](https://github.com/riverwm/river)

OpenGL version/GPU info:
```
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Intel
INFO:     > Renderer: Mesa Intel(R) Xe Graphics (TGL GT2)
INFO:     > Version:  4.6 (Core Profile) Mesa 22.1.1
INFO:     > GLSL:     4.60
```

Screen:
```
Display size: 2256 x 1504
```

### Issue Screenshot

Screenshots are that of `examples/core_basic_window` with varying levels of scaling using `wlr-randr`.

With `wlr-randr --output <output> --scale 1`:
![image](https://user-images.githubusercontent.com/74805176/177482129-9e9cfd2f-09ef-4c91-a84d-2fb804cded7a.png)

With `wlr-randr --output <output> --scale 1.5`:
![image](https://user-images.githubusercontent.com/74805176/177482618-1c395276-c23f-4a3a-8dca-35cf389ce82a.png)

With `wlr-randr --output <output> --scale 0.5`:
![image](https://user-images.githubusercontent.com/74805176/177482997-a3d741bb-e4e9-413a-a352-85acfccd66d1.png)

---

These two screenshots were made by starting the window at a given scale and changing the scale to something else:

`--scale 1` -> `--scale 1.5`:
![image](https://user-images.githubusercontent.com/74805176/177483083-2791c5d8-fe38-4302-b2fa-6423213dc043.png)

`--scale 1.5` -> `--scale 1`:
![image](https://user-images.githubusercontent.com/74805176/177483292-b6e4f0f0-bdd0-48f3-b94b-cafcb32a08e3.png)

Something interesting to note is that the last two instances had a core dump when pressing Escape to close the window:
```
INFO: TEXTURE: [ID 2] Unloaded texture data from VRAM (GPU)
INFO: SHADER: [ID 3] Default shader unloaded successfully
INFO: TEXTURE: [ID 1] Default texture unloaded successfully
core_basic_window: ../src/wayland-client.c:230: wl_proxy_unref: Assertion `proxy->refcount > 0' failed.
Aborted (core dumped)
```

The last two lines replaced what normally would be `INFO: Window closed successfully`.

### Code Example

1. Build Raylib with Wayland support.
2. Use it to try out anything from `examples`.

---

Wayland support was built with the following (as per https://github.com/raysan5/raylib/wiki/CMake-Build-Options):
```
mkdir build && cd build
cmake -DCUSTOMIZE_BUILD=ON -DUSE_WAYLAND=ON ..
make
```

---

Compiling examples:

Commented out a line in `examples/Makefile`:
```diff
diff --git a/examples/Makefile b/examples/Makefile
index 636f5ec6..7f386408 100644
--- a/examples/Makefile
+++ b/examples/Makefile
@@ -313,7 +313,7 @@ ifeq ($(PLATFORM),PLATFORM_DESKTOP)
         LDLIBS = -lraylib -lGL -lm -lpthread -ldl -lrt

         # On X11 requires also below libraries
-        LDLIBS += -lX11
+        #LDLIBS += -lX11
         # NOTE: It seems additional libraries are not required any more, latest GLFW just dlopen them
         #LDLIBS += -lXrandr -lXinerama -lXi -lXxf86vm -lXcursor
```

Built with:
```sh-session
USE_WAYLAND_DISPLAY=TRUE make PLATFORM=PLATFORM_DESKTOP CFLAGS="-I../build/raylib -L../build/raylib"
```

**Comment by raysan5** [comment]
@sagehane I'm afraid this issue is related to windowing system and GLFW. It could be related to automatic framebuffer scaling for HiDPI monitors but it's difficult to track and find a solution that will work fo all use cases.

I will update GLFW as soon as a new version is published, let's see if it solves the issue. In the meantime I'm closing this issue.

**Comment by digitalsignalperson** [comment]
Scaling still broken here in 2023. Arch Linux KDE Plasma /w Wayland, 2560x1600 screen with desktop scaling at 150%. Every raylib demo renders a window with bottom left quadrant with content and the rest empty.

./core_2d_camera
```
INFO: Initializing raylib 4.6-dev
INFO: Supported raylib modules:
INFO:     > rcore:..... loaded (mandatory)
INFO:     > rlgl:...... loaded (mandatory)
INFO:     > rshapes:... loaded (optional)
INFO:     > rtextures:. loaded (optional)
INFO:     > rtext:..... loaded (optional)
INFO:     > rmodels:... loaded (optional)
INFO:     > raudio:.... loaded (optional)
INFO: DISPLAY: Device initialized successfully
INFO:     > Display size: 2560 x 1600
INFO:     > Screen size:  800 x 450
INFO:     > Render size:  800 x 450
INFO:     > Viewport offsets: 0, 0
INFO: GLAD: OpenGL extensions loaded successfully
INFO: GL: Supported extensions count: 228
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Intel
INFO:     > Renderer: Mesa Intel(R) Graphics (RPL-P)
INFO:     > Version:  4.6 (Core Profile) Mesa 23.1.3
INFO:     > GLSL:     4.60
INFO: GL: VAO extension detected, VAO functions loaded successfully
INFO: GL: NPOT textures extension detected, full NPOT textures supported
INFO: GL: DXT compressed textures supported
INFO: GL: ETC2/EAC compressed textures supported
INFO: TEXTURE: [ID 1] Texture loaded successfully (1x1 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 1] Default texture loaded successfully
INFO: SHADER: [ID 1] Vertex shader compiled successfully
INFO: SHADER: [ID 2] Fragment shader compiled successfully
INFO: SHADER: [ID 3] Program shader loaded successfully
INFO: SHADER: [ID 3] Default shader loaded successfully
INFO: RLGL: Render batch vertex buffers loaded successfully in RAM (CPU)
INFO: RLGL: Render batch vertex buffers loaded successfully in VRAM (GPU)
INFO: RLGL: Default OpenGL state initialized successfully
WARNING: GLFW: Error: 65548 Description: Wayland: The platform does not support setting the window position
INFO: TEXTURE: [ID 2] Texture loaded successfully (128x128 | GRAY_ALPHA | 1 mipmaps)
INFO: FONT: Default font loaded successfully (224 glyphs)
INFO: TIMER: Target time per frame: 16.667 milliseconds
INFO: TEXTURE: [ID 2] Unloaded texture data from VRAM (GPU)
INFO: SHADER: [ID 3] Default shader unloaded successfully
INFO: TEXTURE: [ID 1] Default texture unloaded successfully
INFO: Window closed successfully
```

**Comment by raysan5** [comment]
@digitalsignalperson Did you enbale HighDPI flag?

**Comment by digitalsignalperson** [comment]
Thanks, I was not aware of the flag. However, trying it in a few examples I'm not seeing a difference.

- shaders_julia_set.c : uncommented `SetConfigFlags(FLAG_WINDOW_HIGHDPI);` and recompiled but same behaviour as before. 
- shaders_basic_lighting.c : changed to `SetConfigFlags(FLAG_MSAA_4X_HINT | FLAG_WINDOW_HIGHDPI)` and same

**Comment by raysan5** [comment]
@digitalsignalperson Some examples use a RenderTexture, unfortunately in those cases, HighDPI scaling must be managed by users, to do that, raylib provide the function `GetWindowScaleDPI()` to get the scale factor required for HighDPI.

**Comment by higherkinded** [comment]
Necrobumping, but the example in question doesn't use RenderTexture. Experiencing the same without camera/RenderTexture currently.

**Comment by alextopic** [comment]
I recently installed CachyOS Arch Linux / latest KDE Plasma. I also have Ubuntu 24.10 Gnome 47 where it works fine.
I am getting this problem with KDE Plasma running a basic raylib explain. I did make raylib 5.5 GLFW wayland shared libraries.
By default raylib is done with x11. The basic explain runs fine expect this glitch.

Part of my log below:
$ journalctl -r
---
Dec 03 10:21:28 Alextronia kwin_wayland[1132]: kwin_scene_opengl: 0x2: GL_INVALID_OPERATION in glDrawBuffers(unsupported buffer GL_BACK_LEFT)
Dec 03 10:21:28 Alextronia kwin_wayland[1132]: kwin_scene_opengl: 0x2: GL_INVALID_OPERATION in glDrawBuffers(unsupported buffer GL_BACK_LEFT)
Dec 03 10:21:28 Alextronia kwin_wayland[1132]: kwin_scene_opengl: 0x2: GL_INVALID_OPERATION in glDrawBuffers(unsupported buffer GL_BACK_LEFT)
---

**Comment by alextopic** [comment]
Ok, I got it working properly. KDE Plasma you have to go to Control Panel -->Display config

-Make sure monitor set to 100% Scale

-Legcy applications (x11):  Select  "Scaled by the system"   NOT "Apply scaling themselves"

Hope this helps others out, I am amazed it works now.

**Comment by alextopic** [comment]
Well, I did some simple ball bounce Wayland Raylib example that uses the GLFW (Graphics Library Frame Work); which is really the problem and then to support scaling adjustments; beyond raysan's control.

For now, I just make sure in KDE Plasma that I keep it at 100% scale. I also find with multi monitors my example will open on the secondary monitor not the primary monitor; which is very annoying.

**Comment by Kyren223** [comment]
Again, keeping it at 100% is not an option, it changes my other apps scale which makes my PC unusable, the monitor issue is a "feature" of wayland bcz you can't set positions.

It might just be GLFW but I don't think that's the only issue, from testing me and some1 else did, it seems to be some sort of desync/misscommunication between raylib and GLFW

For example, the hack that I was able to find is this:
If you want a 800 by 400 screen for example:
InitWindow with 150% (or whatever scale) so 1200 by 600 in this case
then call SetWindowSize with 800 by 400 and it seems to fix it

There might be other caveats I haven't discovered with this approach
Alternatively build for X11 and run it using xwayland (not ideal, also can have issues)

This should be investigated on raylib's side because there is a pretty high likelihood that this can be fixed, or at least the investigation will make it clear if it's an issue on GLFW's end in which case an issue there can be opened

**Comment by asdromundo** [comment]
After some headaches, I think I have something:
- Setting the FLAG_WINDOW_HIGHDPI before starting the window seems to fix it
```C
SetConfigFlags(FLAG_WINDOW_HIGHDPI);
\\ Before creating the window
InitWindow(screenWidth, screenHeight, "raylib app");
```
- I'm currently **unaware** if you have to _account for the scale_ when getting the **ScreenSize**.
- Stills shows other errors as libdecor, and Wayland not accepting positioning the window

Example in the raylib_game_template
### Before:
![Image](https://github.com/user-attachments/assets/9265473d-37c4-4c8a-a830-6f5587ceb954)

### After:

![Image](https://github.com/user-attachments/assets/fb1ec4a5-b518-4bad-9ec7-d91436c89418)
